File: apis/core-api/internal/adminbootstrap/cognito.go
package adminbootstrap

import (
	"context"
	"errors"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	cip "github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider/types"
)

// CognitoAPI is the subset of the Cognito Identity Provider client this package uses (a seam for
// tests). *cognitoidentityprovider.Client satisfies it.
type CognitoAPI interface {
	AdminCreateUser(ctx context.Context, in *cip.AdminCreateUserInput, optFns ...func(*cip.Options)) (*cip.AdminCreateUserOutput, error)
	AdminGetUser(ctx context.Context, in *cip.AdminGetUserInput, optFns ...func(*cip.Options)) (*cip.AdminGetUserOutput, error)
	AdminEnableUser(ctx context.Context, in *cip.AdminEnableUserInput, optFns ...func(*cip.Options)) (*cip.AdminEnableUserOutput, error)
	AdminAddUserToGroup(ctx context.Context, in *cip.AdminAddUserToGroupInput, optFns ...func(*cip.Options)) (*cip.AdminAddUserToGroupOutput, error)
	AdminDeleteUser(ctx context.Context, in *cip.AdminDeleteUserInput, optFns ...func(*cip.Options)) (*cip.AdminDeleteUserOutput, error)
}

// Cognito is the identity-provider adapter for the back-office pool.
type Cognito struct {
	api    CognitoAPI
	poolID string
}

// NewCognito wires the adapter to a client + the back-office pool id.
func NewCognito(api CognitoAPI, poolID string) *Cognito {
	return &Cognito{api: api, poolID: poolID}
}

// EnsureAdmin creates a PASSWORDLESS, CONFIRMED user (no TemporaryPassword → not
// FORCE_CHANGE_PASSWORD → can EMAIL_OTP immediately — research F1/F2), suppressing the invite
// email, then ensures `admin` group membership. On re-run it reconciles an existing user
// (break-glass — research F4). Returns the immutable `sub` (the token/DB join key — research F3).
func (c *Cognito) EnsureAdmin(ctx context.Context, email, name string) (string, string, error) {
	created, err := c.api.AdminCreateUser(ctx, &cip.AdminCreateUserInput{
		UserPoolId:    aws.String(c.poolID),
		Username:      aws.String(email), // username_attributes=[email] → Cognito generates a UUID username = sub (research F5)
		MessageAction: types.MessageActionTypeSuppress,
		UserAttributes: []types.AttributeType{
			{Name: aws.String("email"), Value: aws.String(email)},
			{Name: aws.String("email_verified"), Value: aws.String("true")},
			{Name: aws.String("name"), Value: aws.String(name)},
		},
		// No TemporaryPassword — required to land the user CONFIRMED on a passwordless pool.
	})

	var (
		outcome  string
		username string
		sub      string
	)

	if err != nil {
		var exists *types.UsernameExistsException
		if !errors.As(err, &exists) {
			return "", "", fmt.Errorf("AdminCreateUser: %w", err)
		}
		// Reconcile an existing user (break-glass).
		outcome = "already-exists"
		got, gerr := c.api.AdminGetUser(ctx, &cip.AdminGetUserInput{
			UserPoolId: aws.String(c.poolID),
			Username:   aws.String(email),
		})
		if gerr != nil {
			return "", "", fmt.Errorf("AdminGetUser: %w", gerr)
		}
		username = aws.ToString(got.Username)
		sub = subFromAttrs(got.UserAttributes)
		if !got.Enabled {
			if _, eerr := c.api.AdminEnableUser(ctx, &cip.AdminEnableUserInput{
				UserPoolId: aws.String(c.poolID),
				Username:   aws.String(username),
			}); eerr != nil {
				return "", "", fmt.Errorf("AdminEnableUser: %w", eerr)
			}
		}
	} else {
		outcome = "created"
		username = aws.ToString(created.User.Username)
		sub = subFromAttrs(created.User.Attributes)
	}

	if sub == "" {
		return "", "", errors.New("cognito: no sub attribute in response")
	}

	// Idempotent: adding an already-member is a no-op (research F4).
	if _, aerr := c.api.AdminAddUserToGroup(ctx, &cip.AdminAddUserToGroupInput{
		UserPoolId: aws.String(c.poolID),
		Username:   aws.String(username),
		GroupName:  aws.String(RoleAdmin),
	}); aerr != nil {
		return "", "", fmt.Errorf("AdminAddUserToGroup: %w", aerr)
	}

	return sub, outcome, nil
}

func subFromAttrs(attrs []types.AttributeType) string {
	for _, a := range attrs {
		if aws.ToString(a.Name) == "sub" {
			return aws.ToString(a.Value)
		}
	}
	return ""
}

// ResolveAdmin looks a user up by email and returns the immutable sub + the real (UUID) username.
// found is false when the user does not exist (idempotent teardown — research G1).
func (c *Cognito) ResolveAdmin(ctx context.Context, email string) (sub, username string, found bool, err error) {
	out, gerr := c.api.AdminGetUser(ctx, &cip.AdminGetUserInput{
		UserPoolId: aws.String(c.poolID),
		Username:   aws.String(email),
	})
	if gerr != nil {
		var nf *types.UserNotFoundException
		if errors.As(gerr, &nf) {
			return "", "", false, nil
		}
		return "", "", false, fmt.Errorf("AdminGetUser: %w", gerr)
	}
	return subFromAttrs(out.UserAttributes), aws.ToString(out.Username), true, nil
}

// DeleteAdmin hard-deletes the user (group memberships vanish with them). deleted is false when the
// user was already gone (idempotent — research G1).
func (c *Cognito) DeleteAdmin(ctx context.Context, username string) (deleted bool, err error) {
	if _, derr := c.api.AdminDeleteUser(ctx, &cip.AdminDeleteUserInput{
		UserPoolId: aws.String(c.poolID),
		Username:   aws.String(username),
	}); derr != nil {
		var nf *types.UserNotFoundException
		if errors.As(derr, &nf) {
			return false, nil
		}
		return false, fmt.Errorf("AdminDeleteUser: %w", derr)
	}
	return true, nil
}

File: apis/edge-api/fleet/src/drivers/cognito.ts
// Driver-pool Cognito Admin adapter (056; moved from apis/edge-api/admin/src/drivers/cognito.ts).
//
// A back-office-authenticated caller provisions identities in the DRIVER pool — an authorized
// server-side provisioning write, NOT cross-pool authentication (049 research R3/R9; the routes
// still carry the back-office authorizer). Mirrors the 006/009 pattern: AdminCreateUser with no
// password, SUPPRESS invite, email_verified — so the driver lands CONFIRMED on a passwordless pool.
//
// ⚠ Unlike shops, the DRIVER pool defines NO RBAC groups (Principle IV) — there is no group step.
// A driver's status and zone live entirely in the platform `public.driver` record.
//
// ⚠ THIS IS NOT A VERBATIM MOVE. The 049 original exported one function, `ensureDriverUser`, which
// swallowed UsernameExistsException, RE-ENABLED a disabled account, and returned the existing sub —
// and the repository layered an `ON CONFLICT (cognito_sub) DO UPDATE` on top of it. Together those
// two "helpful" behaviours meant that creating a driver with an email already in use silently EDITED
// that person's record (name, zone, vehicle) and REACTIVATED the sign-in of someone who had been
// deliberately stood down. It reported success. FR-014 requires a refusal instead, so the
// idempotent-ensure is split into an operation that CREATES and fails loudly on a duplicate, and
// separate operations that enable and disable.
import {
  AdminCreateUserCommand,
  AdminDisableUserCommand,
  AdminEnableUserCommand,
  AdminGetUserCommand,
  type AttributeType,
  CognitoIdentityProviderClient,
  UserNotFoundException,
  UsernameExistsException,
} from "@aws-sdk/client-cognito-identity-provider";

let client: CognitoIdentityProviderClient | undefined;

function getClient(): CognitoIdentityProviderClient {
  if (!client) client = new CognitoIdentityProviderClient({});
  return client;
}

function poolId(): string {
  const id = process.env.DRIVER_USER_POOL_ID;
  if (!id) throw new Error("cognito: DRIVER_USER_POOL_ID is not set");
  return id;
}

function subFromAttrs(attrs: AttributeType[] | undefined): string {
  const sub = attrs?.find((a) => a.Name === "sub")?.Value;
  if (!sub) throw new Error("cognito: no sub attribute in response");
  return sub;
}

/** Raised when the driver pool already holds an account for this address. NOT swallowed. */
export class DriverUserExistsError extends Error {
  constructor(readonly email: string) {
    super(`a driver-pool account already exists for ${email}`);
    this.name = "DriverUserExistsError";
  }
}

/**
 * Create a driver-pool account, returning the stable `sub` (the DB join key).
 *
 * ⚠ NOT idempotent, on purpose. A duplicate address throws `DriverUserExistsError`, which the service
 * turns into FR-014's named 409. The old idempotent form is what let "add a new driver" quietly
 * become "overwrite an existing one".
 */
export async function createDriverUser(email: string, name: string): Promise<string> {
  try {
    const created = await getClient().send(
      new AdminCreateUserCommand({
        UserPoolId: poolId(),
        Username: email,
        MessageAction: "SUPPRESS",
        UserAttributes: [
          { Name: "email", Value: email },
          { Name: "email_verified", Value: "true" },
          { Name: "name", Value: name },
        ],
        // No TemporaryPassword — required to land CONFIRMED on a passwordless pool.
      }),
    );
    return subFromAttrs(created.User?.Attributes);
  } catch (err) {
    if (err instanceof UsernameExistsException) throw new DriverUserExistsError(email);
    throw err;
  }
}

/** Does the driver pool hold an account for this address, and is it enabled? `null` = no account.
 *  Used to compute `accountState` so a half-provisioned driver SHOWS the discrepancy (FR-006). */
export async function lookupDriverUser(
  email: string,
): Promise<{ sub: string; enabled: boolean } | null> {
  try {
    const got = await getClient().send(
      new AdminGetUserCommand({ UserPoolId: poolId(), Username: email }),
    );
    return { sub: subFromAttrs(got.UserAttributes), enabled: got.Enabled !== false };
  } catch (err) {
    if (err instanceof UserNotFoundException) return null;
    throw err;
  }
}

/** Disable the identity account — defence in depth. The record check already refuses a non-active
 *  driver, but disabling means no session can be obtained even if that check were bypassed. */
export async function disableDriverUser(email: string): Promise<void> {
  await getClient().send(new AdminDisableUserCommand({ UserPoolId: poolId(), Username: email }));
}

export async function enableDriverUser(email: string): Promise<void> {
  await getClient().send(new AdminEnableUserCommand({ UserPoolId: poolId(), Username: email }));
}
